api4/export_test.go

```diff
@@ -0,0 +1,213 @@
+		require.False(t, ok)
+	})
+
+	dataDir, found := fileutils.FindDir("data")
+	require.True(t, found)
+	exportDir := filepath.Join(dataDir, *th.App.Config().ExportSettings.Directory)
+
+	th.TestForSystemAdminAndLocal(t, func(t *testing.T, c *model.Client4) {
+		err := os.Mkdir(exportDir, 0700)
+		require.Nil(t, err)
+		defer os.RemoveAll(exportDir)
+		exportName := "export.zip"
+		f, err := os.Create(filepath.Join(exportDir, exportName))
+		require.Nil(t, err)
+		f.Close()
+
+		exports, resp := c.ListExports()
+		require.Nil(t, resp.Error)
+		require.Len(t, exports, 1)
+		require.Equal(t, exports[0], exportName)
+
+		ok, resp := c.DeleteExport(exportName)
+		require.Nil(t, resp.Error)
+		require.True(t, ok)
+
+		exports, resp = c.ListExports()
+		require.Nil(t, resp.Error)
+		require.Empty(t, exports)
+
+		// verify idempotence
+		ok, resp = c.DeleteExport(exportName)
+		require.Nil(t, resp.Error)
+		require.True(t, ok)
+	}, "successfully delete export")
+}
+
+func TestDownloadExport(t *testing.T) {
+	th := Setup(t)
+	defer th.TearDown()
+
+	t.Run("no permissions", func(t *testing.T) {
+		var buf bytes.Buffer
+		n, resp := th.Client.DownloadExport("export.zip", &buf, 0)
+		require.Error(t, resp.Error)
+		require.Equal(t, "api.context.permissions.app_error", resp.Error.Id)
+		require.Zero(t, n)
+	})
+
+	dataDir, found := fileutils.FindDir("data")
+	require.True(t, found)
+	exportDir := filepath.Join(dataDir, *th.App.Config().ExportSettings.Directory)
+
+	th.TestForSystemAdminAndLocal(t, func(t *testing.T, c *model.Client4) {
+		var buf bytes.Buffer
+		n, resp := c.DownloadExport("export.zip", &buf, 0)
+		require.Error(t, resp.Error)
+		require.Equal(t, "api.export.export_not_found.app_error", resp.Error.Id)
+		require.Zero(t, n)
+	}, "not found")
+
+	th.TestForSystemAdminAndLocal(t, func(t *testing.T, c *model.Client4) {
+		err := os.Mkdir(exportDir, 0700)
+		require.Nil(t, err)
+		defer os.RemoveAll(exportDir)
+
+		data := randomBytes(t, 1024*1024)
+		var buf bytes.Buffer
+		exportName := "export.zip"
+		err = ioutil.WriteFile(filepath.Join(exportDir, exportName), data, 0600)
+		require.Nil(t, err)
+
+		n, resp := c.DownloadExport(exportName, &buf, 0)
+		require.Nil(t, resp.Error)
+		require.Equal(t, len(data), int(n))
+		require.Equal(t, data, buf.Bytes())
+	}, "full download")
+
+	th.TestForSystemAdminAndLocal(t, func(t *testing.T, c *model.Client4) {
+		err := os.Mkdir(exportDir, 0700)
+		require.Nil(t, err)
+		defer os.RemoveAll(exportDir)
+
+		data := randomBytes(t, 1024*1024)
+		var buf bytes.Buffer
+		exportName := "export.zip"
+		err = ioutil.WriteFile(filepath.Join(exportDir, exportName), data, 0600)
+		require.Nil(t, err)
+
+		offset := 1024 * 512
+		n, resp := c.DownloadExport(exportName, &buf, int64(offset))
+		require.Nil(t, resp.Error)
+		require.Equal(t, len(data)-offset, int(n))
+		require.Equal(t, data[offset:], buf.Bytes())
+	}, "download with offset")
+}
+
+func BenchmarkDownloadExport(b *testing.B) {
```

Comment:
I've used this mainly to verify that export files are streamed during download and nothing is kept in memory, except for some small buffers.